### ISSUE TITLE
Legion: HDF5 output with a control replicated top-level task

### ISSUE THREAD

**Issue by bandokihiro** [StanfordLegion/legion#907]
Given a `LogicalRegion` `lr` with a `LogicalPartition` `lp`, I would like to dump data to `.h5` without using a `CopyLauncher`. The reason for this is I don't want to allocate space as a field for every output I want to make and instead use a standard `IndexTaskLauncher` to write to file. 

##### What worked on `master`
On `master`, the following worked:
1. in top-level, attach without mapping `AttachLauncher attach_launcher(EXTERNAL_HDF5_FILE, file_lr, file_lr, false, false);` (here, `file_lr` is the logical region used to dump to file)
2. index launch using `lp` and `file_lp` with write privileges for `file_lp`

For some reason that is unclear to me, this approach doesn't work with control replication and nothing is written to file. The instances of the sub-regions of `file_lr` created for the index task don't seem to be tied to the `.h5` file.

##### One file experiments on `control_replication`
To overcome the former issue, I tried to attach in a index launched task `attach_task` (without inline mapping). Since I cannot attach in a non-leaf task, `attach_task` launches itself a leaf task `write_task`. Each sub-region of `file_lp` is virtually mapped in `attach_task` and then attached to the same file `out.h5`. However, they both write to the beginning of the file (no way to provide an offset for instance).

##### Per-shard file experiments on `control_replication`
This approach creates a file for each shard. Now the problem is compaction. The only working (I think) procedure I came up with was to create a new logical region from scratch `local_lr` in `attach_task` of size equal to the volume of the output domain in that shard, attach (without mapping), launch `write_task` with write privileges on `local_lr`, wait for the future and then destroy the logical region/field space/index space associated with `local_lr` still in `attach_task`. If I don't wait for the future, I had some issues. This approach has obviously at least 2 undesirable aspects: the wait and the fact of creating new regions over and over again.

Ideally, I would like to have one output instead of per-shard. But a per-shard solution is acceptable. I don't think the solution I came up with is optimal. 

In summary I have the following questions/comments:
- why does the working approach on `master` (that attaches in top-level) does not work in `control_replication`?
- is there a one file solution for `control_replication` in case I can't attach in top-level? The current problem is that as discussed with @streichler I can't provide an offset and all shards write from the beginning of the dataset
- is there a better per-shard solution that the one I came up with?

**Comment by lightsighter** [comment]
This is one of the bugs that makes #546 a prerequisite for #765. While I consider it a duplicate, I will leave it open for now.

**Comment by bandokihiro** [comment]
Do you expect my solution to induce a memory issue? One of my runs is running out of memory since I implemented that

**Comment by lightsighter** [comment]
@bandokihiro Is this issue resolved now that collective views have merged into the control replication branch (see #546)?

**Comment by bandokihiro** [comment]
The per-shard implementation completely replaced the approach that did not work initially. I am unable to retest this without making a lot of changes... I'll close this issue.
